### PR TITLE
CB-10722 Set AWS root cert ID when creating RDS

### PR DESCRIPTION
**Pull request by lajosrodek**
Contents:
* Extend Redbeams "root certificate registry" entries with cloud provider specific ID ("CloudProviderIdentifier"). This is currently only relevant for AWS, but Azure entries also include the PEM file name without extension as the CloudProviderIdentifier.
* Keep track of the active CloudProviderIdentifier for each newly launched DB server in `DBStack.sslConfig`, and return it in `DatabaseServerV4Response.sslConfig`.
* Provide the desired CloudProviderIdentifier to AWS RDS upon launch. This is a new optional dynamic parameter in `DatabaseServer`. If omitted (the legacy behavior), the RDS will be brought up with whatever the default SSL root cert is for the target account. While it is not possible to query the default SSL root cert directly, the RDS `DescribeCertificates` operation can at least return the list of CloudProviderIdentifier values supported for the target account. (This does not seem to be region-dependent, but the `AmazonRdsClient` is nevertheless instantiated for the target region.)
  * No CloudProviderIdentifier will be prescribed if it is the only one supported as per `DescribeCertificates`. This is not only an optimization but it also avoids some extra time during RDS launch. See [CB-11387](https://jira.cloudera.com/browse/CB-11387) for details.
* After AWS RDS launch completes, its CloudProviderIdentifier will be queried and synced back to `DBStack.sslConfig`. If possible, the version and PEM fields will be also updated.
* Manual tests: local CB, AWS & Azure, with and without DB SSL, DB SSL with and without CloudProviderIdentifier prescription.
* Add some unit test support classes: `AbstractActionTestSupport`, `ExceptionCatcherEventHandlerTestSupport`. These provide means for the invocation of `protected` methods or the usage of `protected` nested types that would be otherwise inaccessible outside the parent package.
* Add new unit tests (JUnit 5 & AssertJ) and amend existing ones. Existing unit tests have been migrated to JUnit 5, and sometimes refactored to use AssertJ as well.

### CODE REVIEW

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java [hunk 1]
```diff
@@ -36,17 +48,25 @@ public ExternalDatabaseStatus getStatus(AuthenticatedContext ac, DatabaseStack d
...
         try {
+            if (preDescribeMessage != null) {
+                LOGGER.debug(preDescribeMessage);
+            }
             describeDBInstancesResult = rdsClient.describeDBInstances(describeDBInstancesRequest);
         } catch (DBInstanceNotFoundException ex) {
-            LOGGER.debug("DB Instance does not exist: {}", ex.getMessage());
-            return ExternalDatabaseStatus.DELETED;
+            LOGGER.debug(notFoundMessage, ex.getMessage());
+            describeDBInstancesResult = null;
```
Comment (doktoric) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java:57]: is this ok to to return with null ?

File: redbeams/src/main/resources/schema/app/20210219153728_CB-10722_Set_AWS_root_cert_ID_when_creating_RDS.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-- // CB-10722 Set AWS root cert ID when creating RDS
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE sslconfig
+    ADD COLUMN IF NOT EXISTS sslcertificateactivecloudprovideridentifier VARCHAR(255);
```
Comment (doktoric) [inline redbeams/src/main/resources/schema/app/20210219153728_CB-10722_Set_AWS_root_cert_ID_when_creating_RDS.sql:5]: do we need to update this field with a default value ?

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java (hunk 1 above)
Comment (lajosrodek) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java:57]: Yes. That was the original behavior of `getDescribeDBInstancesResult()` below (lines 70 -- 89). That method will directly return the response of `getDescribeDBInstancesResultInternal()`. There are two other places where this internal stuff is called, `getStatus()` above (lines 30 -- 38) and the new `getActiveSslRootCertificate()` below (lines 64 -- 72), both handle the `null` response correctly. All mentioned methods have been covered with UT, including cases for the `null` response.

File: redbeams/src/main/resources/schema/app/20210219153728_CB-10722_Set_AWS_root_cert_ID_when_creating_RDS.sql (hunk 2 above)
Comment (lajosrodek) [reply redbeams/src/main/resources/schema/app/20210219153728_CB-10722_Set_AWS_root_cert_ID_when_creating_RDS.sql:5]: No, the code takes care of assuming a default value when needed. Currently, the only place where this happens is `DatabaseServerConfigToDatabaseServerV4ResponseConverter.convert()` that invokes `databaseServerSslCertificateConfig.getLegacyCloudProviderIdentifierByPlatform()` to fetch the default value. Please note that this is cloud platform dependent, just like the `sslcertificateactiveversion` column added earlier in #10068 which is also nullable and resolved using `databaseServerSslCertificateConfig.getLegacyMaxVersionByPlatform()`.

A similar default value logic will be also needed in a future PR that will take of cert rotation for the external DB.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java [hunk 3]
```diff
@@ -48,40 +61,110 @@
...
     }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<AllocateDatabaseServerRequest> event) {
+        AllocateDatabaseServerFailed failure = new AllocateDatabaseServerFailed(resourceId, e);
+        LOGGER.warn("Error launching the database stack:", e);
+        return failure;
+    }
+
+    private void prescribeSslCertificateIfNeeded(CloudConnector<Object> connector, CloudContext cloudContext, CloudCredential cloudCredential, DBStack dbStack,
```
Comment (doktoric) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java:109]: this method is pretty complex. probably some refactor does not really hurt

Comment (lajosrodek) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java:109]: Good point, has been broken up into multiple methods.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/UpdateDatabaseServerRegistrationHandler.java [hunk 4]
```diff
@@ -75,11 +93,96 @@ public void accept(Event<UpdateDatabaseServerRegistrationRequest> event) {
...
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<UpdateDatabaseServerRegistrationRequest> event) {
+        UpdateDatabaseServerRegistrationFailed failure = new UpdateDatabaseServerRegistrationFailed(resourceId, e);
+        LOGGER.warn("Error updating the database server registration:", e);
+        return failure;
+    }
+
+    private void syncSslCertificateIfNeeded(CloudContext cloudContext, CloudCredential cloudCredential, DBStack dbStack, DatabaseStack databaseStack)
```
Comment (biharitomi) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/UpdateDatabaseServerRegistrationHandler.java:116]: The SSL cert syncer logic could be isolated into an individual service.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java [hunk 5]
```diff
@@ -48,40 +61,127 @@
...
```
Comment (biharitomi) [inline redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java:109]: Can we move this prescribe Cert logic into an individual service please?

**Review by biharitomi — CHANGES_REQUESTED** [review]
I loved to see the nice UT coverage, please check my comments

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java [hunk 6]
```diff
@@ -27,6 +28,17 @@
     private AwsClient awsClient;
 
     public ExternalDatabaseStatus getStatus(AuthenticatedContext ac, DatabaseStack dbStack) {
+        DescribeDBInstancesResult describeDBInstancesResult = getDescribeDBInstancesResultInternal(ac, dbStack, null, "DB Instance does not exist: {}");
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java]: Good point, fixed.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java [hunk 7]
```diff
@@ -36,17 +48,25 @@ public ExternalDatabaseStatus getStatus(AuthenticatedContext ac, DatabaseStack d
...
+            LOGGER.debug(notFoundMessage, ex.getMessage());
+            describeDBInstancesResult = null;
         } catch (RuntimeException ex) {
             throw new CloudConnectorException(ex.getMessage(), ex);
         }
+        return describeDBInstancesResult;
+    }
 
-        return describeDBInstancesResult.getDBInstances()
+    public CloudDatabaseServerSslCertificate getActiveSslRootCertificate(AuthenticatedContext ac, DatabaseStack dbStack) {
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsRdsStatusLookupService.java:65]: Fixed.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java [hunk 8]
```diff
@@ -48,40 +61,127 @@
...
+    }
+
+    private void prescribeSslCertificateAws(CloudConnector<Object> connector, CloudContext cloudContext, CloudCredential cloudCredential,
+            DatabaseServer databaseServer, String cloudPlatform, String desiredSslCertificateIdentifier) {
+        Set<String> availableSslCertificateIdentifiers = getAvailableSslCertificateIdentifiers(connector, cloudContext, cloudCredential, cloudPlatform);
+        if (availableSslCertificateIdentifiers.contains(desiredSslCertificateIdentifier)) {
+            // A more proper condition would be to check if desiredSslCertificateIdentifier was the default among availableSslCertificateIdentifiers,
+            // but this cannot be directly decided without launching a DB server and inspecting its associated SSL root cert.
+            if (availableSslCertificateIdentifiers.size() == 1) {
+                LOGGER.debug("Desired SSL certificate CloudProviderIdentifier is the default for cloud platform \"{}\": \"{}\". " +
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java]: Fixed in new class `DatabaseServerSslCertificatePrescriptionService`.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/UpdateDatabaseServerRegistrationHandler.java (hunk 4 above)
Comment (lajosrodek) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/UpdateDatabaseServerRegistrationHandler.java]: Good point, has been moved to the new class `DatabaseServerSslCertificateSyncService` and covered in `DatabaseServerSslCertificateSyncServiceTest`.

File: redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java (hunk 5 above)
Comment (lajosrodek) [reply redbeams/src/main/java/com/sequenceiq/redbeams/flow/redbeams/provision/handler/AllocateDatabaseServerHandler.java]: Good point, has been moved to the new class `DatabaseServerSslCertificatePrescriptionService` and covered in `DatabaseServerSslCertificatePrescriptionServiceTest`.

**Review by biharitomi — APPROVED** [review]
LGTM

**Pull request merged**